src/commonTest/kotlin/co/remotectrl/ctrl/event/CtrlAggregatePlayerTest.kt

```diff
@@ -8,218 +8,373 @@ import kotlin.test.assertTrue
+        assertAggregateEventResult(
+            expectedAggregateEvent,
+            expectedAggregate = expectedAggregate,
+            played.result
+        )
     }
 
-    @Test
-    fun given_Player_when_Event_played_then_apply_changes() {
-        val player = getPlayer()
-
-        val played = player.playForEvent(
-            aggregateEvent = CtrlAggregateEvent(
-                eventId = CtrlEventId("1"),
-                aggregateId = CtrlId("1"),
-                version = 1,
-                mutableEvent = StubChangedEvent()
-            )
+    private fun getAggregate(changeVal: Int, latestVersion: Int = 0) = CtrlAggregate(
+        id = CtrlId("1"),
+        latestVersion = latestVersion,
+        mutable = StubMutable(
+            changeVal = changeVal
         )
+    )
 
-        assertTrue(played is CtrlTry.Success)
+    @Test
+    fun given_Player_when_Event_played_then_apply_changes() {
+        val player = CtrlAggregatePlayer(aggregate = getAggregate(0))
 
-        assertEquals("1", played.result.id.value)
-        assertEquals(1, played.result.latestVersion)
-        assertEquals(1, played.result.mutable.changeVal)
+        val expectedAggregateEvent = getAggregateEvent("1", "1", 1, StubChangedEvent())
+        val played = player.play(
+            aggregateEvent = expectedAggregateEvent
+        )
 
-        assertEquals("1", player.aggregate.id.value)
-        assertEquals(1, player.aggregate.latestVersion)
-        assertEquals(1, player.aggregate.mutable.changeVal)
+        val expectedAggregate = getAggregate(1, 1)
+        assertPlayed(
+            expectedAggregateEvent = expectedAggregateEvent,
+            expectedAggregate = expectedAggregate,
+            played
+        )
+        assertPlayer(
+            expectedAggregate = expectedAggregate,
+            player
+        )
     }
 
     @Test
     fun given_Player_when_Event_with_different_aggregate_id_played_then_failure() {
-        val player = getPlayer()
-
-        val played = player.playForEvent(
-            aggregateEvent = CtrlAggregateEvent(
-                eventId = CtrlEventId("1"),
-                aggregateId = CtrlId("2"), //note the aggregate id loaded in the player
-                version = 1,
-                mutableEvent = StubChangedEvent()
-            )
+        val expectedAggregateEvent = getAggregateEvent("1", "2", 1, StubChangedEvent())
+
+        val expectedAggregate = getAggregate(0)
+        val player = CtrlAggregatePlayer(aggregate = expectedAggregate)
+        val played = player.play(
+            aggregateEvent = expectedAggregateEvent
+        )
+
+
+        assertPlayedFailedData<AggregateEventInvalidAggregateIdCause<StubMutable>>(
+            expectedAggregateEvent = expectedAggregateEvent,
+            expectedAggregate = expectedAggregate,
+            played
+        )
+        assertPlayer(
+            expectedAggregate = expectedAggregate,
+            player
         )
+    }
 
+    private inline fun <
```

Comment:
This is a more stylistic nit. But it feels like these helper functions can be moved to a test harness/helper. Leaving this class to be just test cases. These potentially could eve be used down stream by users of the library.

I also started thinking about DSL helpers, and starting to type it out. But the amount of code was not simpler. I found in simplifying a recent test suite, that reducing the common bits in the test, helped present the business logic. This made it easier for others to grok the test

This feels noisy and takes away user freedom from a test
```kotlin
fun sampleTestSuccess = withPlayer(getAggregate("0"), TestPlayer.Success) {
}
```

This is pretty succinct and doesn't require much optimiaztion.
```kotlin
fun sampleTestSuccess() {
   val expectedAggregate = getAggregate(0)
   val player = CtrlAggregatePlayer(aggregate = expectedAggregate)
}
```
Centralizing the logic.
```kotlin
fun SampleTestSuccess() {
  val (expectedAggregate, player) = createAggregateForTest(0)
}
```

I think the third is the most streamlined, and eases tests. If anything ever changes with player creation, or aggregate, it's centralized in one helper methods. This is being pedantic all around though, as it flows pretty well. But with these helpers/dsl this could aid in providing a downstream testing helpers for consumers.

There is the difference of a `CtrlAggregatePlayerTest` or a CtrlMutablePlayerTest`. I handled this type of variation with traits. It was pretty easily digested by users later, they just amended the needed traits to their test class.

```
class MyPlayerTest : ICtrlAggregatePlayerHelper {}
class MyMutablePlayerTest : ICtrlMutablePlayerHelper{}
```

Where that interface could potentially implement the above i.e.

```kotlin
interface IEventCtrlTestHelper  {
   abstract fun createPlayer(aggregate: CtrlAggregate): PlayerType 
   fun createAggregateForTest(id: Int=0) : Pair<CtrlAggregate, PlayerType> {
      val aggregate = CtrlAggregate(...)
      val player = createPlayer(aggregate)
   }
}
interface ICtrlAggregatePlayerHelper  : IEventCtrlTestHelper   {
   fun createPlayer(aggregate: CtrlAggregate)  = CtrlAggregatePlayer(aggregate = aggregate) 
}
```

Reply:
I completely agree. but because it's in a test lib I don't know how to package. gradle makes this complicated